File: 0001-twoSum/two-sum.ts
// 1st try

function twoSum(nums: number[], target: number): number[] {
    let arrObj: { [key: number]: number } = {};
    for (let i = 0; i < nums.length; i++) {
        const expected = target - nums[i];
        if (arrObj.hasOwnProperty(expected)) {
            return [arrObj[expected], i];
        }
        arrObj[nums[i]] = i;
    }
    return [];
}


// better
function twoSum2(nums: number[], target: number): number[] {
    const arrObj: Record<number, number> = {}
    for (let i = 0; i < nums.length; i++) {
        const expected = target - nums[i];
        if (expected in arrObj) {
            return [arrObj[expected], i]
        }
        arrObj[nums[i]] = i
    }

    return []
}

File: 0001-twoSum/two-sum.go
func twoSum(nums []int, target int) []int {
	arrMap := make(map[int]int)

	for i, num := range nums {
		expected := target - num

		if idx, exist := arrMap[expected]; exist {
			return []int{idx, i}
		}

		arrMap[num] = i
	}
	return []int{}
}
